packages/core/src/lib/BufferedLog.ts

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2020 Google Inc. All Rights Reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+import {Log} from './Log';
+
+enum LogLevel {
+  Debug,
+  Info,
+  Warn,
+  Error,
+}
+
+/** Data class to store level and message of pending */
+class PendingLog {
+  constructor(public level: LogLevel, public message: string) {}
+}
+
+/**
+ * A Log that wraps another Log, saving up all the log calls to be applied when flush is called.
+ *
+ * It doesn't currently support arguments to the log messages.
+ */
+export class BufferedLog implements Log {
+  private pendingLogs: Array<PendingLog> = [];
+
+  constructor(private innerLog: Log) {}
+
+  /* eslint-disable no-invalid-this */
+  debug = this.addLogFunction(LogLevel.Debug);
+  info = this.addLogFunction(LogLevel.Info);
+  warn = this.addLogFunction(LogLevel.Warn);
+  error = this.addLogFunction(LogLevel.Error);
+  /* eslint-enable no-invalid-this */
+
+  /** Creates a function that adds a log at the given level. */
+  private addLogFunction(level: LogLevel): (message: string) => void {
+    return (message: string): void => {
+      this.pendingLogs.push(new PendingLog(level, message));
+    };
+  }
```

Comment:
```suggestion
  private pendingLogs: Array<PendingLog> = [];
  public debug: (message: string) => void;
  public info: (message: string) => void;
  public warn: (message: string) => void;
  public error: (message: string) => void;

  constructor(private innerLog: Log) {
    this.debug = this.addLogFunction(LogLevel.Debug);
    this.info = this.addLogFunction(LogLevel.Info);
    this.warn = this.addLogFunction(LogLevel.Warn);
    this.error = this.addLogFunction(LogLevel.Error);
  }

  /** Creates a function that adds a log at the given level. */
  private addLogFunction(level: LogLevel): (message: string) => void {
    return (message: string): void => {
      this.pendingLogs.push(new PendingLog(level, message));
    };
  }
```

Reply:
Yeah, that's a good suggestion.

So as far as I can tell (purely just because it compiles and seems to work), what I've written is valid code. So this means that the lint check is wrong. In which case is it better to work around the faulty lint tool or to just disable it temporarily so we can carry on?

(Or maybe there is something actually wrong with my code that I don't understand.)

Reply:
Because I'm curious to know if the problem is with eslint or with my code, I [filed an issue with eslint](https://github.com/eslint/eslint/issues/13894).

Reply:
FWIW, this is what the transpiled code looks like:
```typescript
class BufferedLog {
    constructor(innerLog) {
        this.innerLog = innerLog;
        this.pendingLogs = [];
        /* eslint-disable no-invalid-this */
        this.debug = this.addLogFunction(LogLevel.Debug);
        this.info = this.addLogFunction(LogLevel.Info);
        this.warn = this.addLogFunction(LogLevel.Warn);
        this.error = this.addLogFunction(LogLevel.Error);
        // This is the code inside the constructor.
        console.log('Where does the constructor code go?');
    }
    /* eslint-enable no-invalid-this */
    /** Creates a function that adds a log at the given level. */
    addLogFunction(level) {
        return (message) => {
            this.pendingLogs.push(new PendingLog(level, message));
        };
    }
```
I was wondering if that code could lead to some kind of undefined behaviour, but doesn't look like the case. Would you mind adding a link to the issue in the comment and a TO-DO to checkout out the issue response in the future? Besides that, LGTM then.